### PR TITLE
Added more geometry-relevant navigation methods

### PR DESCRIPTION
**Pull request by DManstrator**
## Pull Request

- [x] I have checked for similar PRs.
- [x] I have read the [contributing guidelines](https://github.com/4Soft-de/vec-model/blob/develop/.github/CONTRIBUTING.md).

### Changes

- [x] Code
- [x] Documentation
- [ ] Other: 

### Description

Added 17 new navigation methods for the `VecDocumentVersion` (`DocumentVersionNavs`), `VecOnWayPlacement` (`PlacementNavs`) and `VecSpecification` / `HasSpecifications` (`SpecificationNavs`).

The navigation methods are about
- placeable element roles,
- placements,
- geometry nodes,
- view items,
- building blocks,
- positionings,
- locations,
- components and
- geometry segments.

### Breaking changes

This includes a breaking change chaning a signature of the `SpecificationNavs`:
```diff
- public static Function<HasSpecifications<VecSpecification>, Stream<VecOccurrenceOrUsage>> allOccurrenceOrUsages()
+ public static Function<HasSpecifications<VecSpecification>, List<VecOccurrenceOrUsage>> allOccurrenceOrUsages()
```

### CODE REVIEW

File: v12x/src/main/java/com/foursoft/harness/vec/v12x/navigations/DocumentVersionNavs.java [hunk 1]
```diff
@@ -109,4 +110,118 @@ private DocumentVersionNavs() {
                 .collect(Collectors.toList());
     }
 
+    public static Function<VecDocumentVersion, VecPlaceableElementRole> placeableElementRoleBy(
+            final String specificationValue,
+            final String occurrenceOrUsageId) {
+        return dv -> SpecificationNavs.componentsBy(specificationValue).apply(dv)
```
Comment (takebene) [inline v12x/src/main/java/com/foursoft/harness/vec/v12x/navigations/DocumentVersionNavs.java:116]: dv  -> documentVersion

File: v12x/src/main/java/com/foursoft/harness/vec/v12x/navigations/DocumentVersionNavs.java [hunk 2]
```diff
@@ -109,4 +110,118 @@ private DocumentVersionNavs() {
...
+                .map(VecPartOccurrence::getRoles)
+                .flatMap(Collection::stream)
+                .filter(VecPlaceableElementRole.class::isInstance)
+                .map(VecPlaceableElementRole.class::cast)
+                .collect(StreamUtils.findOne());
+    }
+
+    public static Function<VecDocumentVersion, VecPlacement> placementBy(final String specificationValue,
+                                                                         final String occurrenceOrUsageId) {
+        return dv -> {
```
Comment (takebene) [inline v12x/src/main/java/com/foursoft/harness/vec/v12x/navigations/DocumentVersionNavs.java:128]: dv -> documentVersion

File: v12x/src/main/java/com/foursoft/harness/vec/v12x/navigations/DocumentVersionNavs.java [hunk 3]
```diff
@@ -109,4 +110,118 @@ private DocumentVersionNavs() {
...
+            return dv.getSpecificationsWithType(VecPlacementSpecification.class).stream()
+                    .map(VecPlacementSpecification::getPlacements)
+                    .flatMap(Collection::stream)
+                    .filter(p -> p.getPlacedElement().stream().anyMatch(r -> r.equals(role)))
+                    .collect(StreamUtils.findOne());
+        };
+    }
+
+    public static Function<VecDocumentVersion, VecGeometryNode2D> geometryNode2DBy(final VecNodeLocation location) {
+        return dv -> {
```
Comment (takebene) [inline v12x/src/main/java/com/foursoft/harness/vec/v12x/navigations/DocumentVersionNavs.java:141]: same

File: v12x/src/main/java/com/foursoft/harness/vec/v12x/navigations/SpecificationNavs.java [hunk 4]
```diff
@@ -73,4 +64,69 @@ private SpecificationNavs() {
         };
     }
 
+    public static Function<HasSpecifications<VecSpecification>, Stream<VecOccurrenceOrUsage>> allOccurrenceOrUsages() {
```
Comment (takebene) [inline v12x/src/main/java/com/foursoft/harness/vec/v12x/navigations/SpecificationNavs.java:67]: Aren't this just PartUsages?

Comment (takebene) [reply v12x/src/main/java/com/foursoft/harness/vec/v12x/navigations/SpecificationNavs.java:67]: Stream?

File: v12x/src/main/java/com/foursoft/harness/vec/v12x/navigations/SpecificationNavs.java [hunk 5]
```diff
@@ -73,4 +64,69 @@ private SpecificationNavs() {
...
+    }
+
+    /**
+     * Gets the {@link VecCompositionSpecification}s and gets
+     * their {@link VecCompositionSpecification#getComponents() components}.
+     *
+     * @return A possibly-empty list of Components.
+     */
+    public static Function<HasSpecifications<VecSpecification>, List<VecPartOccurrence>> components() {
+        return hasSpecifications -> hasSpecifications
```
Comment (takebene) [inline v12x/src/main/java/com/foursoft/harness/vec/v12x/navigations/SpecificationNavs.java:83]: hasSpecifications -> specification

Comment (takebene) [reply v12x/src/main/java/com/foursoft/harness/vec/v12x/navigations/SpecificationNavs.java:83]: or pass the interface HasSpecifications

File: v12x/src/main/java/com/foursoft/harness/vec/v12x/navigations/SpecificationNavs.java (hunk 4 above)
Comment (DManstrator) [reply v12x/src/main/java/com/foursoft/harness/vec/v12x/navigations/SpecificationNavs.java:67]: > Aren't this just PartUsages?

No, `DocumentVersion` also implements that. This method was used before for `DocumentVersion`s only but then I saw that the method is defined in the `HasSpecifications` interface so I used it instead.

> Stream?

Good catch. I used that before in another navigation method and extracted that, but I forgot to adjust the type correctly.

File: v12x/src/main/java/com/foursoft/harness/vec/v12x/navigations/SpecificationNavs.java (hunk 5 above)
Comment (DManstrator) [reply v12x/src/main/java/com/foursoft/harness/vec/v12x/navigations/SpecificationNavs.java:83]: I don't see how I could pass the Interface to that method, then it wouldn't be a `Function` anymore. I believe the rename is just fine.

File: v12x/src/main/java/com/foursoft/harness/vec/v12x/navigations/SpecificationNavs.java (hunk 4 above)
Comment (DManstrator) [reply v12x/src/main/java/com/foursoft/harness/vec/v12x/navigations/SpecificationNavs.java]: @takebene: Btw. the change from `Stream<VecOccurrenceOrUsage>` to `List<VecOccurrenceOrUsage>` is a breaking change. The method already existed before but was moved within the class. This should be a minor release update then, shouldn't it?

**Review by takebene — APPROVED** [review]

**Pull request merged**